src/libraries/Common/src/System/Net/NTAuthentication.Managed.cs

```diff
@@ -0,0 +1,919 @@
+                            {
+                                mechListWriter.WriteObjectIdentifier(NtlmOid);
+                            }
+
+                            _spnegoMechList = mechListWriter.Encode();
+                            mechListWriter.CopyTo(writer);
+                        }
+
+                        using (writer.PushSequence(new Asn1Tag(TagClass.ContextSpecific, (int)NegTokenInit.MechToken)))
+                        {
+                            writer.WriteOctetString(ntlmNegotiateMessage);
+                        }
+                    }
+                }
+            }
+
+            return writer.Encode();
+        }
+
+        private unsafe byte[] ProcessSpNegoChallenge(byte[] challenge)
+        {
+            NegState state = NegState.Unknown;
+            string? mech = null;
+            byte[]? blob = null;
+
+            AsnReader reader = new AsnReader(challenge, AsnEncodingRules.DER);
+            AsnReader challengeReader = reader.ReadSequence(new Asn1Tag(TagClass.ContextSpecific, (int)NegotiationToken.NegTokenResp));
+            reader.ThrowIfNotEmpty();
+
+            // NegTokenResp::= SEQUENCE {
+            //    negState[0] ENUMERATED {
+            //        accept - completed(0),
+            //        accept - incomplete(1),
+            //        reject(2),
+            //        request - mic(3)
+            //    } OPTIONAL,
+            // --REQUIRED in the first reply from the target
+            //    supportedMech[1] MechType OPTIONAL,
+            // --present only in the first reply from the target
+            // responseToken[2] OCTET STRING  OPTIONAL,
+            // mechListMIC[3] OCTET STRING  OPTIONAL,
+            // ...
+            // }
+
+            challengeReader = challengeReader.ReadSequence();
+            while (challengeReader.HasData)
+            {
+                Asn1Tag tag = challengeReader.PeekTag();
+                if (tag.TagClass == TagClass.ContextSpecific)
+                {
+                    NegTokenResp dataType = (NegTokenResp)tag.TagValue;
+                    AsnReader specificValue = new AsnReader(challengeReader.PeekContentBytes(), AsnEncodingRules.DER);
+
+                    switch (dataType)
+                    {
+                        case NegTokenResp.NegState:
+                            state = specificValue.ReadEnumeratedValue<NegState>();
+                            specificValue.ThrowIfNotEmpty();
+                            break;
+                        case NegTokenResp.SupportedMech:
+                            mech = specificValue.ReadObjectIdentifier();
+                            specificValue.ThrowIfNotEmpty();
+                            break;
+                        case NegTokenResp.ResponseToken:
+                            blob = specificValue.ReadOctetString();
+                            specificValue.ThrowIfNotEmpty();
+                            break;
+                        default:
```

Comment:
This loop+switch will decode multiples and out-of-order values.  Let's not add dynamic behavior where we don't want it, and just read linearly.

Reply:
Fixed.